File: epi_judge_java_solutions/epi/SortList.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class SortList {
  @EpiTest(testDataFile = "sort_list.tsv")

  public static ListNode<Integer> stableSortList(ListNode<Integer> L) {

    // Base cases: L is empty or a single node, nothing to do.
    if (L == null || L.next == null) {
      return L;
    }

    // Find the midpoint of L using a slow and a fast pointer.
    ListNode<Integer> preSlow = null, slow = L, fast = L;
    while (fast != null && fast.next != null) {
      preSlow = slow;
      fast = fast.next.next;
      slow = slow.next;
    }

    if (preSlow != null) {
      preSlow.next = null; // Splits the list into two equal-sized lists.
    }
    return SortedListsMerge.mergeTwoSortedLists(stableSortList(L),
                                                stableSortList(slow));
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SortList.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python_solutions/sort_list.py
from typing import Optional

from list_node import ListNode
from sorted_lists_merge import merge_two_sorted_lists
from test_framework import generic_test


def stable_sort_list(L: ListNode) -> Optional[ListNode]:

    # Base cases: L is empty or a single node, nothing to do.
    if L is None or L.next is None:
        return L

    # Find the midpoint of L using a slow and a fast pointer.
    pre_slow, slow, fast = None, L, L
    while fast and fast.next:
        pre_slow = slow
        fast, slow = fast.next.next, slow.next

    if pre_slow:
        pre_slow.next = None  # Splits the list into two equal-sized lists.
    return merge_two_sorted_lists(stable_sort_list(L), stable_sort_list(slow))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('sort_list.py', 'sort_list.tsv',
                                       stable_sort_list))
